File: Python/528__Random_Pick_with_Weight.py
class Solution(object):

    def __init__(self, w):
        """
        :type w: List[int]
        """
        import random
        self.array = []
        self.total = sum(w)
        print(w)
        for i, num in enumerate(w):
            self.array.append(num if i == 0 else self.array[-1]+num)
        print(self.array)

    def pickIndex(self):
        """
        :rtype: int
        """
        i, j = 0, len(self.array)-1
        rand_num = random.randint(1,self.total)
        while i <= j:
            mid = (i+j)/2
            if self.array[mid] < rand_num:
                i = mid + 1
            else:
                j = mid - 1
        return i



# Your Solution object will be instantiated and called as such:
# obj = Solution(w)
# param_1 = obj.pickIndex()

File: Java/528__Random_Pick_with_Weight.java
class Solution {

    List<Integer> sums = new ArrayList<>();
    int sum = 0;
    Random rand = new Random();

    public Solution(int[] w) {
        for (int val : w) {
            sum += val;
            sums.add(sum);
        }
    }

    public int pickIndex() {
        int val = rand.nextInt(sum);
        int l = 0, r = sums.size() - 1;
        while (l < r) {
            int mid = (l + r) / 2;
            if (val >= sums.get(mid))
                l = mid + 1;
            else
                r = mid;
        }
        return l;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(w);
 * int param_1 = obj.pickIndex();
 */
